### PR TITLE
feature(dspy): Add MyScale in Retrieve

### PR DESCRIPTION
**Pull request by usamajamil43**
Add MyScaleRM to support retrieval for MyScaleDB:

- MyScaleDB: https://github.com/myscale/myscaledb

### CODE REVIEW

File: dspy/retrieve/MyScaleRM.py
Comment (arnavsinghvi11) [inline dspy/retrieve/MyScaleRM.py:116]: provide the installation command here

File: dspy/retrieve/MyScaleRM.py
Comment (arnavsinghvi11) [inline dspy/retrieve/MyScaleRM.py:163]: can these requests be cached? could you adapt from [DSPy's caching functionalities](https://github.com/stanfordnlp/dspy/blob/f2a96137288925be7a9ad46b5c40c1e0425bd531/dsp/modules/cache_utils.py#L9)?

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

**Comment by arnavsinghvi11** [comment]
Hi @usamajamil43, thanks for the contribution! left a few comments in the PR. Noticed some todos so feel free to mark the PR as ready to merge once the changes are made. 

Couple of additional changes:
1. can you run `ruff check . --fix-only` to fix the failing test?
2. could you add documentation for MyScaleRM to the Retrieval Model Clients API? Feel free to follow some of the [example RM documentation](https://github.com/stanfordnlp/dspy/blob/f2a96137288925be7a9ad46b5c40c1e0425bd531/docs/api/retrieval_model_clients/ChromadbRM.md) for reference.

File: dspy/retrieve/MyScaleRM.py
Comment (usamajamil43) [reply dspy/retrieve/MyScaleRM.py:116]: I have added installation commands in the exception.

**Comment by usamajamil43** [comment]
Hey @arnavsinghvi11 , I have fixed the issues and added the documentation. Kindly, take a look

File: dspy/retrieve/MyScaleRM.py
Comment (arnavsinghvi11) [inline dspy/retrieve/MyScaleRM.py:129]: this is close, but I was thinking something along the lines of what's done in [dspy.OpenAI](https://github.com/stanfordnlp/dspy/blob/f2a96137288925be7a9ad46b5c40c1e0425bd531/dsp/modules/gpt3.py#L218) for openai chat and completions requests. Since MyScale relies on OpenAI embeddings, it would be good to enable caching functionality for it.

A longer term goal would be to have the OpenAI/custom embeddings function you've added as abstract utils functions outside of MyScale. Feel free to actually add a TODO comment so we can track this!

File: dspy/retrieve/MyScaleRM.py
Comment (arnavsinghvi11) [inline dspy/retrieve/MyScaleRM.py:129]: this is close, but I was thinking something along the lines of what's done in [dspy.OpenAI](https://github.com/stanfordnlp/dspy/blob/f2a96137288925be7a9ad46b5c40c1e0425bd531/dsp/modules/gpt3.py#L218) for openai chat and completions requests. Since MyScale relies on OpenAI embeddings, it would be good to enable caching functionality for it.

A longer term goal would be to have the OpenAI/custom embeddings function you've added as abstract utils functions outside of MyScale. Feel free to actually add a TODO comment so we can track this!

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

File: dspy/retrieve/MyScaleRM.py
Comment (usamajamil43) [reply dspy/retrieve/MyScaleRM.py:129]: What about this way?
```python
@CacheMemory.cache
@functools.lru_cache(maxsize=50)  # Cache the most recent 100 queries
def get_embeddings(self, queries: List[str]) -> List[List[float]]:
      return self._get_embeddings(queries)
```

Comment (arnavsinghvi11) [reply dspy/retrieve/MyScaleRM.py:129]: I was thinking something like this:

```
@CacheMemory.cache
def cached_embeddings_request(**kwargs):
    return openai.embeddings.create(**kwargs)


@functools.lru_cache(maxsize=None if cache_turn_on else 0)
@NotebookCacheMemory.cache
def cached_embeddings_request(**kwargs):
    return cached_embeddings_request(**kwargs)
```

and then you can configure the embeddings as needed. (might need a separate method for the local model embeddings but a similar methodology should work)!

Comment (usamajamil43) [reply dspy/retrieve/MyScaleRM.py:129]: Sure!

Comment (usamajamil43) [reply dspy/retrieve/MyScaleRM.py]: Hi @arnavsinghvi11, I have added a multi-layer Cace as you have requested.

**Comment by arnavsinghvi11** [comment]
Thanks @usamajamil43 ! LGTM. Unfortunately, I cannot test the caches for this myself, but could you verify that with the existing behavior before pushing? Also, the PR has a few merge conflicts to resolve but should be ready to merge following that!

**Comment by arnavsinghvi11** [comment]
Hi @usamajamil43 , could you run `ruff check . --fix-only` and commit again? Ready to merge after that!

**Comment by usamajamil43** [comment]
Hi, @arnavsinghvi11 Done!

**Comment by arnavsinghvi11** [comment]
Hi @usamajamil43 , can you rebase to the latest version on main? should fix the failing ruff test and ready to merge.

**Comment by usamajamil43** [comment]
Hey @arnavsinghvi11,

I've tried to resolve the conflicts, but I'm still getting conflicts in files like lm.py and evaluate.py, which I haven't worked on. I'm not sure which changes to keep and which to discard in these files.

**Comment by arnavsinghvi11** [comment]
Hi @usamajamil43 , I believe if you rebase to the latest commit on the main branch and then run ruff, this should fix the existing error. I tried doing this by pulling your branch locally which resolved it, but I don't have access to push changes.

**Comment by usamajamil43** [comment]
@arnavsinghvi11  Done!

**Comment by usamajamil43** [comment]
@arnavsinghvi11  Can you please take a look? I was busy, so @ahsansaeed878 did the merge.

**Comment by arnavsinghvi11** [comment]
Thanks @usamajamil43 and @ahsansaeed878 for your patience with the PR! merged.

**Pull request merged**
